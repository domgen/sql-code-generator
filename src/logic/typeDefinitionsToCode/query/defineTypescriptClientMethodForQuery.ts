import { camelCase } from 'change-case';
import { castQueryNameToTypescriptTypeName } from '../common/castQueryNameToTypescriptTypeName';

/**
 * given:
 * - typescript type definition of input and output
 *
 * produce:
 * - typescript function that gets the inputs and returns promise of output
 * - typescript function that calls database in best practice way
 *   - uses yesql to pass inputs into parameterized statement
 *   - parses result and extracts output
 *   - logs inputs and output stats
 */
export const defineTypescriptClientMethodForQuery = ({
  name,
  relativePathToQueryExport,
}: {
  name: string;
  relativePathToQueryExport: string;
}) => {
  const typescriptTypeName = castQueryNameToTypescriptTypeName({ name });
  const inputTypeName = `${typescriptTypeName}Input`;
  const outputTypeName = `${typescriptTypeName}Output`;
  const typescriptClientMethodName = camelCase(typescriptTypeName); // i.e., same as typename, except camel case
  const clientMethodDefinition = `
import { mysql as prepare } from 'yesql';
import { query } from '${relativePathToQueryExport}';
import { ${inputTypeName}, ${outputTypeName} } from '../types';
import { DatabaseExecuteCommand, LogMethod } from './common';

export const ${typescriptClientMethodName} = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: ${inputTypeName};
}): Promise<${outputTypeName}[]> => {
  // 1. define the query with yesql
  const { sql: preparedSql, values: preparedValues } = prepare(query)(input);

  // 2. log that we're running the request
  logDebug('${typescriptClientMethodName}.input', { input });

  // 3. execute the query
  const output = await dbExecute({ sql: preparedSql, values: preparedValues });

  // 4. log that we've executed the request
  logDebug('${typescriptClientMethodName}.output', { output });

  // 5. return the output
  return output;
};
  `.trim();
  return clientMethodDefinition;
};
