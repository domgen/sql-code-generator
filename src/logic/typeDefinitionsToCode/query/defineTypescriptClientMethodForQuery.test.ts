import { defineTypescriptClientMethodForQuery } from './defineTypescriptClientMethodForQuery';

describe('defineTypescriptClientMethodForQuery', () => {
  it('should define the method accurately based on name and import path', () => {
    const code = defineTypescriptClientMethodForQuery({
      name: 'upsert_image',
      relativePathToQueryExport: '../image/upsert',
    });
    console.log(code);
    expect(code).toMatchSnapshot();
  });
});
