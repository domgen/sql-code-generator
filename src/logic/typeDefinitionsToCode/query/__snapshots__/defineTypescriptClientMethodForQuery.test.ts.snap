// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineTypescriptClientMethodForQuery should define the method accurately based on name and import path 1`] = `
"import { mysql as prepare } from 'yesql';
import { query } from '../image/upsert';
import { SqlQueryUpsertImageInput, SqlQueryUpsertImageOutput } from '../types';
import { DatabaseExecuteCommand, LogMethod } from './common';

export const sqlQueryUpsertImage = async ({
  dbExecute,
  logDebug,
  input,
}: {
  dbExecute: DatabaseExecuteCommand;
  logDebug: LogMethod;
  input: SqlQueryUpsertImageInput;
}): Promise<SqlQueryUpsertImageOutput[]> => {
  // 1. define the query with yesql
  const { sql: preparedSql, values: preparedValues } = prepare(query)(input);

  // 2. log that we're running the request
  logDebug('sqlQueryUpsertImage.input', { input });

  // 3. execute the query
  const output = await dbExecute({ sql: preparedSql, values: preparedValues });

  // 4. log that we've executed the request
  logDebug('sqlQueryUpsertImage.output', { output });

  // 5. return the output
  return output;
};"
`;
